### PR TITLE
Store: Add wc-api error handler for data-layer

### PR DESCRIPTION
**Pull request by coderkevin**
Note: The ethos of this PR is explained in the README in #21274

This injects an error handler into the data layer WooCommerce API calls
by adding a specialized `dispatchRequest` function to be used instead of
the normal one. It has the ability to track error codes and fire actions
based on them.

This improves on our existing error state by giving a centralized place that can monitor the status of a given WooCommerce site via the API calls that are sent to it. As of now, there are four possible states: `uninitialized` (before any API calls are made), `available` (after a successful API call), `unavailable` (after a 404 from the site), and `unknown_error` (for all other error states). More states may be added after their nature is determined.

Note: Only products requests are using this as of now. It seemed like a good first step. Other requests can be ported after this is working.

To Test: (works best on a site with multiple pages of products)

1. `npm run test-client client/extensions/woocommerce/state/wc-api` and ensure tests pass.
2. `npm start`
3. Visit `http://calypso.localhost:3000/store/products/<site url>`
4. Open JS console and Redux DevTools and drill down to `extensions.woocommerce.wcApi.siteData[ <site id> ].status. It should be 'available'.
5. Deactivate your WooCommerce plugin on your site via wp-admin on another tab.
6. Select another page of products. A `WOOCOMMERCE_WC_API_UNAVAILABLE` action should be present, and the status in state should now be `unavailable`.
7. Reactivate WooCommerce and select another page. `WOOCOMMERCE_WC_API_SUCCESS` should be present, and status should be `available` again.
8. Disconnect from your network or otherwise disrupt your connection to the WPCOM API. Select another page. `WOOCOMMERCE_WC_API_UNKNOWN_ERROR` should be present, and status should be `unknown_error`.
9. Reconnect to network and select another page. All should be successful again.

### CODE REVIEW

File: client/extensions/woocommerce/state/wc-api/utils.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+					wcApiSite: action.siteId,
+					action,
+					error: {
+						status: data.status,
+						code: data.body.code,
+						message: data.body.message,
+					},
+				} );
+
+				// TODO: Log to logstash
```
Comment (timmyc) [inline client/extensions/woocommerce/state/wc-api/utils.js]: I also think we should bump some mc stats here with the error code, and if possible, endpoint.

Comment (timmyc) [reply client/extensions/woocommerce/state/wc-api/utils.js]: But could be a follow-up PR.

**Comment by timmyc** [comment]
I think this will be a great utility to have for fully understanding the scope of how often timeouts are being hit across client sites: ref p8yzl4-1dd-p2

File: client/extensions/woocommerce/state/wc-api/utils.js (hunk 1 above)
Comment (coderkevin) [reply client/extensions/woocommerce/state/wc-api/utils.js]: Stats could make sense, but we might be able to get stats out of kibana as well.

**Comment by coderkevin** [comment]
Thanks, Timmy!

**Comment by coderkevin** [comment]
Rebased, even though there were no conflicts, just because it's been sitting a while.

**Pull request merged**